File: cad/worker/mcad_worker/framing.py
"""LSP-style length-prefixed framing for the Go-Python bridge.

Frame format (per design §3):
    Content-Length: NNNN\r\n
    Content-Type: application/json; charset=utf-8\r\n
    \r\n
    <NNNN bytes of body>

Only Content-Length is mandatory; Content-Type is included on write and
accepted-but-ignored on read. Any other header is silently skipped.
On parse failure a FramingError is raised — the caller should treat this
as a fatal protocol error and exit cleanly.
"""

from __future__ import annotations

import io


class FramingError(Exception):
    """Raised when the framing layer encounters a malformed header or EOF."""


def read_frame(stream: io.RawIOBase) -> bytes:
    """Read one framed message from *stream* and return the raw body bytes.

    Raises:
        FramingError: if headers are malformed, Content-Length is missing,
                      or the stream closes before delivering the full body.
    """
    headers: dict[str, str] = {}

    # Read header lines until the blank separator line.
    while True:
        line_bytes = _read_line(stream)
        if line_bytes is None:
            raise FramingError("Unexpected EOF while reading headers")
        line = line_bytes.decode("ascii").rstrip("\r\n")
        if line == "":
            # Blank line — end of headers.
            break
        if ":" not in line:
            raise FramingError(f"Malformed header line (no colon): {line!r}")
        key, _, value = line.partition(":")
        headers[key.strip().lower()] = value.strip()

    if "content-length" not in headers:
        raise FramingError("Missing Content-Length header")

    try:
        length = int(headers["content-length"])
    except ValueError:
        raise FramingError(
            f"Content-Length is not an integer: {headers['content-length']!r}"
        )

    if length < 0:
        raise FramingError(f"Content-Length is negative: {length}")

    # Read the exact body.
    body = _read_exactly(stream, length)
    if body is None or len(body) < length:
        raise FramingError(
            f"Unexpected EOF: expected {length} bytes, got {len(body) if body else 0}"
        )
    return body


def write_frame(stream: io.RawIOBase, body: bytes) -> None:
    """Write *body* as a framed message to *stream* and flush.

    Args:
        stream: A writable binary stream (e.g. sys.stdout.buffer).
        body:   Raw UTF-8 JSON bytes to send.
    """
    header = (
        f"Content-Length: {len(body)}\r\n"
        f"Content-Type: application/json; charset=utf-8\r\n"
        f"\r\n"
    ).encode("ascii")
    stream.write(header + body)
    stream.flush()


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _read_line(stream: io.RawIOBase) -> bytes | None:
    """Read one CRLF- or LF-terminated line from *stream*.

    Returns None on immediate EOF (before any bytes of the line are read).
    The terminator is included in the returned bytes.
    """
    buf = bytearray()
    while True:
        ch = stream.read(1)
        if not ch:
            return None if not buf else bytes(buf)
        buf += ch
        if ch == b"\n":
            return bytes(buf)


def _read_exactly(stream: io.RawIOBase, n: int) -> bytes | None:
    """Read exactly *n* bytes from *stream*.

    Returns the bytes read (may be fewer than *n* on EOF) or None if n == 0.
    """
    if n == 0:
        return b""
    chunks: list[bytes] = []
    remaining = n
    while remaining > 0:
        chunk = stream.read(remaining)
        if not chunk:
            break
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)

File: cad/internal/bridge/framing.go
// Package bridge implements the length-prefixed JSON framing protocol used
// between the Go MCP server and the Python CAD worker subprocess.
//
// Frame format (design §3):
//
//	Content-Length: NNNN\r\n
//	Content-Type: application/json; charset=utf-8\r\n
//	\r\n
//	<body bytes>
package bridge

import (
	"bufio"
	"fmt"
	"io"
	"strconv"
	"strings"
)

const (
	headerContentLength = "Content-Length"
	headerContentType   = "Content-Type"
	contentTypeJSON     = "application/json; charset=utf-8"
)

// WriteFrame writes a single length-prefixed frame to w.
// It emits both Content-Length and Content-Type headers followed by CRLF CRLF
// and then the raw body bytes.
func WriteFrame(w io.Writer, body []byte) error {
	header := fmt.Sprintf(
		"%s: %d\r\n%s: %s\r\n\r\n",
		headerContentLength, len(body),
		headerContentType, contentTypeJSON,
	)
	if _, err := io.WriteString(w, header); err != nil {
		return fmt.Errorf("bridge.WriteFrame: write header: %w", err)
	}
	if _, err := w.Write(body); err != nil {
		return fmt.Errorf("bridge.WriteFrame: write body: %w", err)
	}
	return nil
}

// ReadFrame reads one length-prefixed frame from r.
// It reads CRLF-terminated header lines until the blank separator, extracts
// the Content-Length value, then reads exactly that many bytes as the body.
//
// On any header parse failure the caller must drain the reader and restart
// the worker (design §3 "On header parse failure").
func ReadFrame(r *bufio.Reader) ([]byte, error) {
	contentLength := -1

	for {
		line, err := r.ReadString('\n')
		if err != nil {
			return nil, fmt.Errorf("bridge.ReadFrame: read header line: %w", err)
		}
		// Strip trailing CRLF or LF.
		line = strings.TrimRight(line, "\r\n")

		// Blank line == end of headers.
		if line == "" {
			break
		}

		parts := strings.SplitN(line, ": ", 2)
		if len(parts) != 2 {
			return nil, fmt.Errorf("bridge.ReadFrame: malformed header: %q", line)
		}
		name := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])

		if strings.EqualFold(name, headerContentLength) {
			n, parseErr := strconv.Atoi(value)
			if parseErr != nil || n < 0 {
				return nil, fmt.Errorf("bridge.ReadFrame: invalid Content-Length %q: %w", value, parseErr)
			}
			contentLength = n
		}
		// Other headers (e.g. Content-Type) are accepted and ignored.
	}

	if contentLength < 0 {
		return nil, fmt.Errorf("bridge.ReadFrame: missing Content-Length header")
	}

	body := make([]byte, contentLength)
	if _, err := io.ReadFull(r, body); err != nil {
		return nil, fmt.Errorf("bridge.ReadFrame: read body (%d bytes): %w", contentLength, err)
	}
	return body, nil
}
